### PR TITLE
Added CSquareMatrixCM to the toolboxes

### PR DESCRIPTION
**Pull request by vdweide**
## Proposed Changes
As agreed during today's developer's meeting, this PR adds a general square matrix utility to the toolboxes such that it can be used by other people. It is not used in the current source tree of the develop branch and therefore it should not have any affect on current simulations.
 

## Related Work
N/A

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X ] I am submitting my contribution to the develop branch.
- [ X] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [ X] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

**Review by pcarruscag — COMMENTED** [review]
Sorry it took me a while to have a look at this, couple questions below.
I still want to add some math syntax sugar to this.

File: Common/src/toolboxes/CGeneralSquareMatrixCM.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+  dgetri_(&sz, mat.data(), &sz, ipiv.data(), work.data(), &sz, &info);
+  if(info != 0) SU2_MPI::Error(string("Matrix inversion failed"), CURRENT_FUNCTION);
+
+#else
+
+  /*--- Standard implementation without Lapack. Create a local augmented
+        matrix to carry out the actual inversion. ---*/
+  const int M = Size();
+  ColMajorMatrix<passivedouble> augmentedmatrix(M, 2*M);
```
[the remark this answers is not in the document]
Comment (vdweide) [reply Common/src/toolboxes/CGeneralSquareMatrixCM.cpp:78]: I just took what I had in an older source and didn't look any further. However, we should definitely have one implementation for one task, so we should choose one of the implementation. I don't have a preference. This means though that there should be an additional layer in between C2DContainer and CSymmetricMatrix/CGeneralSquareMatrixCM to make sure that both CSymmetricMatrix and CGeneralSquareMatrixCM have access to it. Putting the computation of the inverse in C2DContainer works, but that is an ugly solution in my opinion,

File: Common/src/toolboxes/CGeneralSquareMatrixCM.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    passivedouble alpha = 1.0, beta = 0.0;
+    char trans = 'N';
+
+    dgemm_(&trans, &trans, &M, &N, &M, &alpha, mat.data(), &M,
+           mat_in.data(), &M, &beta, mat_out.data(), &M);
+#else
+    /*--- Naive product. ---*/
+    for (int i = 0; i < M; ++i) {
+      for (int j = 0; j < N; ++j) {
+        mat_out(i,j) = 0.0;
```
[the remark this answers is not in the document]
Comment (vdweide) [reply Common/src/toolboxes/CGeneralSquareMatrixCM.cpp:158]: That's correct. The reason why I didn't do this is that for good performance you have to use OpenBLAS or MKL anyways. But we can certainly improve this implementation, if that is desired.

File: Common/src/toolboxes/CGeneralSquareMatrixCM.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pcarruscag) [reply Common/src/toolboxes/CGeneralSquareMatrixCM.cpp:78]: It does not look like the two algorithms have any row-major - col-major advantage, I say we go with LU decomposition as a free function (perhaps part of the blas functions) as that gives more options when it comes to solve linear systems.

We should take the opportunity and move the Eigenvalue routines that exist now in CNumerics @jayantmukho

**Comment by vdweide** [comment]
Should we close this one or are people still interested in this functionality?

**Comment by pcarruscag** [comment]
Keep it open, I'll try to make the changes we talked about.

**Review by pcarruscag — APPROVED** [review]

**Pull request merged**